### PR TITLE
Linting with rubocop.

### PR DESCRIPTION
**Pull request by iamAzeem**
- Incorporated Robucop changes.
- Added `.rubocop.yml` configuration file.
- Updated README.
- Updated C++ tests:
  - Added error-handling.
  - Added deallocations for dynamic allocations wherever applicable.
- Tested locally with:
  - Ruby: `ruby 2.5.1p57 (2018-03-29 revision 63029) [x86_64-linux-gnu]`
  - Rubocop: `1.8.1`

Signed-off-by: Azeem Sajid <azeem.sajid@gmail.com>

### CODE REVIEW

**Comment by iamAzeem** [comment]
Right now, only the following linter issues remain:

```shell
$ rubocop
Inspecting 65 files
..CC.............C..CC.....C.CCC..C.....CC...........C.CC....C.CC

Offenses:

Rakefile:26:1: C: Style/GlobalVars: Do not introduce global variables.
$gemspec = Bundler::GemHelper.gemspec
^^^^^^^^
ext/extconf_common.rb:7:36: C: [Correctable] Style/RegexpLiteral: Use %r around regular expression.
vendor = ENV.fetch('GEM_HOME', '')[/^[^ ]*\/vendor\//]
                                   ^^^^^^^^^^^^^^^^^^
ext/extconf_common.rb:21:1: C: [Correctable] Style/IfUnlessModifier: Favor modifier if usage when having a single-line body. Another good alternative is the usage of control flow &&/||.
if ENV.key?('TAGLIB_DIR') && !File.directory?(ENV['TAGLIB_DIR'])
^^
ext/extconf_common.rb:32:8: C: Style/GlobalVars: Do not introduce global variables.
unless $LDFLAGS.split(' ').include?('-static-libstdc++')
       ^^^^^^^^
ext/extconf_common.rb:33:46: C: [Correctable] Style/SoleNestedConditional: Consider merging nested conditions into outer unless conditions.
  error 'You must have libstdc++ installed.' unless have_library('stdc++')
                                             ^^^^^^
ext/extconf_common.rb:47:1: C: Style/GlobalVars: Do not introduce global variables.
$CFLAGS << ' -DSWIG_TYPE_TABLE=taglib'
^^^^^^^
lib/taglib/base.rb:6:3: C: Style/Documentation: Missing top-level module documentation comment.
  module FileOpenable
  ^^^^^^
lib/taglib/id3v2.rb:7:5: C: Style/Documentation: Missing top-level class documentation comment.
    class Frame
    ^^^^^
lib/taglib/id3v2.rb:18:9: C: [Correctable] Style/StringLiterals: Prefer single-quoted strings when you don't need string interpolation or special symbols.
        "#<%s:0x%x>" % [self.class.to_s, object_id << 1]
        ^^^^^^^^^^^^
lib/taglib/id3v2.rb:18:12: C: Style/FormatStringToken: Prefer annotated tokens (like %<foo>s) over unannotated tokens (like %s).
        "#<%s:0x%x>" % [self.class.to_s, object_id << 1]
           ^^
lib/taglib/id3v2.rb:18:17: C: Style/FormatStringToken: Prefer annotated tokens (like %<foo>s) over unannotated tokens (like %s).
        "#<%s:0x%x>" % [self.class.to_s, object_id << 1]
                ^^
lib/taglib/id3v2.rb:18:22: C: [Correctable] Style/FormatString: Favor format over String#%.
        "#<%s:0x%x>" % [self.class.to_s, object_id << 1]
                     ^
lib/taglib/mp4.rb:14:3: C: Style/Documentation: Missing top-level class documentation comment.
  class Item
  ^^^^^
lib/taglib/mp4.rb:23:3: C: Style/Documentation: Missing top-level class documentation comment.
  class ItemMap
  ^^^^^
taglib-ruby.gemspec:1:1: C: Gemspec/RequiredRubyVersion: required_ruby_version should be specified.
# frozen-string-literal: true
^
tasks/ext.rake:6:1: C: Style/GlobalVars: Do not introduce global variables.
$plat = ENV['PLATFORM'] || 'i386-mingw32'
^^^^^
tasks/ext.rake:12:22: C: Style/GlobalVars: Do not introduce global variables.
tmp_arch = "#{tmp}/#{$plat}"
                     ^^^^^
tasks/ext.rake:17:1: C: Style/GlobalVars: Do not introduce global variables.
$cross_config_options = ["--with-opt-dir=#{install_dir}"]
^^^^^^^^^^^^^^^^^^^^^
tasks/ext.rake:30:24: C: Style/GlobalVars: Do not introduce global variables.
  ext.cross_platform = $plat
                       ^^^^^
tasks/ext.rake:31:35: C: Style/GlobalVars: Do not introduce global variables.
  ext.cross_config_options.concat($cross_config_options)
                                  ^^^^^^^^^^^^^^^^^^^^^
tasks/ext.rake:38:40: C: Style/GlobalVars: Do not introduce global variables.
Rake::ExtensionTask.new('taglib_base', $gemspec) do |ext|
                                       ^^^^^^^^
tasks/ext.rake:41:40: C: Style/GlobalVars: Do not introduce global variables.
Rake::ExtensionTask.new('taglib_mpeg', $gemspec) do |ext|
                                       ^^^^^^^^
tasks/ext.rake:44:41: C: Style/GlobalVars: Do not introduce global variables.
Rake::ExtensionTask.new('taglib_id3v1', $gemspec) do |ext|
                                        ^^^^^^^^
tasks/ext.rake:47:41: C: Style/GlobalVars: Do not introduce global variables.
Rake::ExtensionTask.new('taglib_id3v2', $gemspec) do |ext|
                                        ^^^^^^^^
tasks/ext.rake:50:39: C: Style/GlobalVars: Do not introduce global variables.
Rake::ExtensionTask.new('taglib_ogg', $gemspec) do |ext|
                                      ^^^^^^^^
tasks/ext.rake:53:42: C: Style/GlobalVars: Do not introduce global variables.
Rake::ExtensionTask.new('taglib_vorbis', $gemspec) do |ext|
                                         ^^^^^^^^
tasks/ext.rake:56:48: C: Style/GlobalVars: Do not introduce global variables.
Rake::ExtensionTask.new('taglib_flac_picture', $gemspec) do |ext|
                                               ^^^^^^^^
tasks/ext.rake:59:40: C: Style/GlobalVars: Do not introduce global variables.
Rake::ExtensionTask.new('taglib_flac', $gemspec) do |ext|
                                       ^^^^^^^^
tasks/ext.rake:62:39: C: Style/GlobalVars: Do not introduce global variables.
Rake::ExtensionTask.new('taglib_mp4', $gemspec) do |ext|
                                      ^^^^^^^^
tasks/ext.rake:65:40: C: Style/GlobalVars: Do not introduce global variables.
Rake::ExtensionTask.new('taglib_aiff', $gemspec) do |ext|
                                       ^^^^^^^^
tasks/ext.rake:68:39: C: Style/GlobalVars: Do not introduce global variables.
Rake::ExtensionTask.new('taglib_wav', $gemspec) do |ext|
                                      ^^^^^^^^
tasks/ext.rake:113:20: C: Security/Open: The use of Kernel#open is a serious security risk.
    IO.copy_stream(open(taglib_url), f)
                   ^^^^
tasks/gemspec_check.rake:7:19: C: Style/GlobalVars: Do not introduce global variables.
  gemspec_files = $gemspec.files
                  ^^^^^^^^
tasks/swig.rake:19:5: C: [Correctable] Style/IfUnlessModifier: Favor modifier unless usage when having a single-line body. Another good alternative is the usage of control flow &&/||.
    unless File.directory?(ENV['TAGLIB_DIR'])
    ^^^^^^
test/aiff_file_write_test.rb:52:52: C: [Correctable] Style/SymbolProc: Pass &:read as an argument to open instead of a block.
      picture_data = File.open(PICTURE_FILE, 'rb') { |f| f.read }
                                                   ^^^^^^^^^^^^^^
test/flac_file_test.rb:87:71: C: [Correctable] Style/SymbolProc: Pass &:read as an argument to open instead of a block.
        picture_data = File.open('test/data/globe_east_90.jpg', 'rb') { |f| f.read }
                                                                      ^^^^^^^^^^^^^^
test/flac_file_write_test.rb:44:52: C: [Correctable] Style/SymbolProc: Pass &:read as an argument to open instead of a block.
      picture_data = File.open(PICTURE_FILE, 'rb') { |f| f.read }
                                                   ^^^^^^^^^^^^^^
test/id3v2_write_test.rb:93:54: C: [Correctable] Style/SymbolProc: Pass &:read as an argument to open instead of a block.
        picture_data = File.open(PICTURE_FILE, 'rb') { |f| f.read }
                                                     ^^^^^^^^^^^^^^
test/mp4_file_write_test.rb:40:44: C: [Correctable] Style/SymbolProc: Pass &:read as an argument to open instead of a block.
      data = File.open(PICTURE_FILE, 'rb') { |f| f.read }
                                           ^^^^^^^^^^^^^^
test/mp4_items_test.rb:36:38: C: [Correctable] Style/PreferredHashMethods: Use Hash#key? instead of Hash#has_key?.
        assert_equal true, @item_map.has_key?("#{ITUNES_LEADER}too")
                                     ^^^^^^^^
test/vorbis_tag_test.rb:104:71: C: [Correctable] Style/SymbolProc: Pass &:read as an argument to open instead of a block.
        picture_data = File.open('test/data/globe_east_90.jpg', 'rb') { |f| f.read }
                                                                      ^^^^^^^^^^^^^^
test/wav_file_test.rb:69:57: C: [Correctable] Style/SymbolProc: Pass &:read as an argument to open instead of a block.
          @picture_data = File.open(PICTURE_FILE, 'rb') { |f| f.read }
                                                        ^^^^^^^^^^^^^^
test/wav_file_write_test.rb:52:52: C: [Correctable] Style/SymbolProc: Pass &:read as an argument to open instead of a block.
      picture_data = File.open(PICTURE_FILE, 'rb') { |f| f.read }
                                                   ^^^^^^^^^^^^^^

65 files inspected, 43 offenses detected, 15 offenses auto-correctable
```

For future contributions, a thorough linter review may also be incorporated (locally or via CI, or both) in the code review process to keep the codebase as clean as possible.

File: README.md [hunk 1]
```diff
@@ -34,107 +32,138 @@ with header files (and a C++ compiler of course):
 
 Then do:
 
-    gem install taglib-ruby
+```shell
```
Comment (robinst) [inline README.md:35]: Not really a fan of the changes in this file. Why does Rubocop care about this file?

**Review by robinst — COMMENTED** [review]
I personally don't think these changes are very important, but don't object to making them either. Just one inline question, but otherwise LGTM.

File: README.md (hunk 1 above)
Comment (iamAzeem) [reply README.md:35]: Actually, some of these changes were due to the markdownlint extension in VSCode.

**Comment by iamAzeem** [comment]
@robinst: I totally understand that. I think that these changes can be postponed to a later major release.

File: README.md (hunk 1 above)
Comment (robinst) [reply README.md:35]: I would undo the `shell` changes in this file, other than that we can merge this PR.

Comment (iamAzeem) [reply README.md]: Removed `shell` with https://github.com/robinst/taglib-ruby/pull/108/commits/a64d623bd8c87b32f7e63e0287425b43cd874194.

**Comment by robinst** [comment]
Thanks for making the changes, merged now.

**Pull request merged**